configure cmd for junit

stevekuznetsov opened the pull request:
This pull does the following:
 * sets up a jUnit output text file artifact
 * gets tests using `os::cmd` to output to it
 * sets up test suite and test case bookends
 * inserts test suite markings into `test/cmd/*.sh`
 * formats tests to better break things up

@deads2k @liggitt @soltysh PTAL - I suggest to look at the first commit only to see changes, or look at the entire changes with whitespace disabled.

soltysh commented:
LGTM

soltysh commented:
Although I'm still thinking about those indentations... ;)

deads2k commented:
> Although I'm still thinking about those indentations... ;)

@stevekuznetsov I think I'm against indenting like that.  I don't think it enhances the readability of the script.

liggitt commented:
> @stevekuznetsov I think I'm against indenting like that. I don't think it enhances the readability of the script.

Agree. I wouldn't expect more than one package per test-cmd script, honestly.

deads2k commented:
> Agree. I wouldn't expect more than one package per test-cmd script, honestly.

Eh, I'm ok with multiple packages.  We sort of did that with "X set of tests: ok".

deads2k commented on hack/cmd_util.sh:
```diff
@@ -182,16 +184,18 @@ function os::cmd::internal::expect_exit_code_run_grep() {
 
 	if (( cmd_succeeded && test_succeeded )); then
-		os::text::print_green "SUCCESS after ${time_elapsed}s: ${name}"
+		os::text::print_green "SUCCESS after ${time_elapsed}s: ${name}" | tee "{JUNIT_REPORT_OUTPUT}"
```
I would have expected `tee "${JUNIT_REPORT_OUTPUT}"`.  Is the dollar sign not needed for some reason?

deads2k replied:
Someone is going to call this method without `JUNIT_RESPORT_OUTPUT` set.  Make this command work as expected for them and add a test to the tests for this file to prove that it works.

stevekuznetsov replied:
> Make this command work as expected

Without that variable set right now:
 * no file is written
 * no error is raised
 * nothing fails to run

What else is expected?

deads2k commented on test/cmd/admin.sh:
```diff
@@ -29,273 +30,304 @@ os::log::install_errexit
-os::cmd::expect_success_and_not_text "oc get rolebindings/cluster-admin         --output-version=v1 --template='{{.subjects}}' -n default" 'cascaded-group'
-os::cmd::expect_success_and_not_text "oc get scc/restricted                     --output-version=v1 --template='{{.groups}}'"              'cascaded-group'
-echo "user-group-cascade: ok"
+os::test::declare_package_start "github.com/openshift/origin/test/cmd/admin"
```
Every test we have is under "github.com/openshift/origin".  Why would I explicitly declare it?  I'd expect this to be as flat as possible and the exceptions would be explicitly called out.

deads2k commented on hack/junit.sh:
```diff
@@ -0,0 +1,23 @@
+}
+
+# os::test::declare_test_end prints a message declaring the end of a test case
+function os::test::declare_test_end() {
```
Keep track of the starts and ends.  I'd like to be able to check depth during `test-cmd.sh` to make sure no one forgot one and to be sure that `test_end` isn't called if we haven't started.  This will help keep us tidy and matching.

deads2k commented:
A few comments.  I largely like the change.

stevekuznetsov commented on hack/test-cmd.sh:
```diff
@@ -55,6 +56,11 @@ tests=( $(find_tests ${1:-.*}) )
 
+if [[ -z "${JUNIT_REPORT_OUTPUT+x}" ]]; then
+  # if no jUnit report output file is set, throw away our jUnit info
+  JUNIT_REPORT_OUTPUT="/dev/null"
```
@deads2k the jUnit functions need `$JUNIT_REPORT_OUTPUT` to be set. I don't think it's reasonable to stick a check in every single function, so instead an `if` case like this in any script using jUnit is more appropriate.

stevekuznetsov commented:
@deads2k ready for another look

deads2k commented:
> Without that variable set right now:
> 
> no file is written
> no error is raised
> nothing fails to run
> What else is expected?

You're setting it in `test-cmd`, but every single place that calls these methods will need to do that same detection and set to `/dev/null`.  That's unreasonable.

deads2k commented:
[test]

deads2k commented on hack/cmd_util.sh:
```diff
@@ -332,21 +348,21 @@ function os::cmd::internal::print_try_until_results() {
 # this is used to make the try_until_* output more concise
 function os::cmd::internal::mark_attempt() {
-	echo -e '\x1e' >> "${os_cmd_internal_tmpout}" | tee "${os_cmd_internal_tmperr}"
+	echo -e '\x1e' >> "${os_cmd_internal_tmpout}"
```
Why did this change?

deads2k commented on hack/cmd_util.sh:
```diff
@@ -156,9 +169,10 @@ function os::cmd::internal::expect_exit_code_run_grep() {
 	local test_eval_func=${4:-os::cmd::internal::success_func}
 
 	os::cmd::internal::init_tempdir
+	os::test::junit::declare_test_start
```
If I nest `expect_success` under a `try_until_success`, will you correctly collapse the repeated test in the junit output and report success for it?

stevekuznetsov replied:
You're saying `os::cmd::try_until_success 'os::cmd::expect_success'` ? That' not legal.

deads2k commented on hack/lib/test/junit.sh:
```diff
@@ -0,0 +1,112 @@
+    local suite_name=$1
+    local num_suites=${NUM_SUITES_IN_FLIGHT:-0}
+    
+    echo "==== BEGIN TEST SUITE github.com/openshift/origin/test/${suite_name} =====" >> "${JUNIT_REPORT_OUTPUT}"
```
The fact that `github.com/openshift/origin/test` is hardcoded suggests that its a zero entropy statement and therefore worthless.  Why do we have it?

stevekuznetsov replied:
Nested test suites in jUnit for Origin will be rooted at `github.com/openshift/origin`. This allows these tests to nest correctly in that schema.

deads2k commented on hack/lib/test/junit.sh:
```diff
@@ -0,0 +1,112 @@
+    
+    echo "==== BEGIN TEST SUITE github.com/openshift/origin/test/${suite_name} =====" >> "${JUNIT_REPORT_OUTPUT}"
+    NUM_SUITES_IN_FLIGHT=$(( ${num_suites} + 1 ))
+    export NUM_SUITES_IN_FLIGHT
```
namespace this.

stevekuznetsov commented:
re[test]

soltysh commented:
It definitely reads better without those indents. Keeping my LGTM.

deads2k commented:
failed the spot twice.  Looks broken.

stevekuznetsov commented:
re[test]

stevekuznetsov commented:
@deads2k we have green

deads2k commented on hack/cmd_util.sh:
```diff
@@ -155,10 +168,14 @@ function os::cmd::internal::expect_exit_code_run_grep() {
 	# default expected test code to 0 for success
 	local test_eval_func=${4:-os::cmd::internal::success_func}
 
+	local -a junit_log
```
spare me reading, what does a `-a` do here?

stevekuznetsov replied:
declares it as an array

deads2k commented on hack/cmd_util.sh:
```diff
@@ -181,19 +198,31 @@ function os::cmd::internal::expect_exit_code_run_grep() {
+
+	junit_log+=( "$(os::cmd::internal::print_results)" )
+	# append inside of a subshell so that IFS doesn't get propagated out
+	( IFS=$'\n'; echo "${junit_log[*]}" >> "${JUNIT_REPORT_OUTPUT:-/dev/null}" )
```
I hate bash so much.

stevekuznetsov replied:
Once the traps PR makes it in, there will be a nice `os::util::join` function that we can use and forget about the demonic sacrifice that goes on behind the scenes to make it happen

deads2k commented on hack/cmd_util.sh:
```diff
@@ -314,6 +343,8 @@ function os::cmd::internal::specific_code_func() {
 	local expected_code=$1
 	local exit_code=$2
 
+	echo "expecting ${expected_code} got ${exit_code}" >> /tmp/test.out
```
Intentional output?

deads2k commented:
I don't see any output of the junit text file or the resulting XML for test-cmd.

stevekuznetsov commented:
@deads2k no, this pull request, as the title of the commit suggests, adds test suite and test case markers for jUnit parsing. The parser needs to be configured to consume whatever we output as a result of this pull. This pull was originally intended to be a quick shoo-in that just added markers... two weeks later, I can see that was naive. At that time, I was planning on doing the `junitreport` and `vagrant-openshift` pieces after this.

deads2k commented:
I should see at least the text file even if nothing parses it, but isn't your parser already in?

stevekuznetsov commented:
@deads2k Parser is in, specific regex for parsing this output is not, as the output never existed before and was needed to be invented. I'll get this pull to start generating the output file to artifacts.

deads2k commented:
@stevekuznetsov tag me by name when this is ready

stevekuznetsov commented:
adding `junitreport` bits as requested @deads2k now [wip], more integration test cases to add

stevekuznetsov commented:
@deads2k parser bits are done. we now generate the logfile to parse into XML. seeing as `test-cmd` fails immediately when it hits an error, the best way to get XML to be produced is to add that as a handler func trapped on error. the jUnit output itself is going to need a trap on error as well, to close up any nested suites that are in flight when everything crumbles. once the trapping PR is in, I'll add those features to a new pull. PTAL

stevekuznetsov commented:
@deads2k @Miciah this is ready for a final look.

Today, I rebased it and updated the suite markers for new tests, added logic to generate the XML report on exit from `test-cmd` and did some minor changes as evidenced by the last three commits.

stevekuznetsov commented:
blocked on https://github.com/openshift/origin/pull/8001

stevekuznetsov commented:
@deads2k prereq merged, left are the vital commits:
 - `configure test-cmd to emit parsable output for jUnit` : this adds support to `os::cmd` for jUnit-parseable output (Bash)
 - `refactored nested suites builder` : this makes a necessary but tangential refactor for `junitreport` (Go)
 - `added support for os::cmd parsing to junitreport` : extends `junitreport` to handle the `os::cmd` output (Go)
 - `added jUnit XML publishing step to exit trap` : hooks up `JUNIT_REPORT=true` to actually generating a report (Bash)

PTAL

stevekuznetsov commented:
re[test]

stevekuznetsov commented:
@deads2k added suite statements to `test/end-to-end/core.sh` so that the jUnit bits didn't error, once we figure out a good way of deconflicting the names of the tests in that suite (which are conflicting due to being called in a function) we get jUnit for e2e basically for free. I'll leave that for a follow-up

stevekuznetsov commented:
@deads2k this is ready for a look

soltysh commented on hack/test-cmd.sh:
```diff
@@ -188,6 +210,22 @@ atomic-enterprise start \
+os::util::sed '5,10d' ${BASETMPDIR}/node-config-broken.yaml
+os::cmd::expect_failure_and_text "openshift ex validate node-config ${BASETMPDIR}/node-config-broken.yaml" 'ERROR'
+echo "validation: ok"
+os::test::junit::declare_suite_end
```
Since in some places we do, others we don't `echo "<group test>: ok"`, how about adding those output as part of `os::test::junit::declare_suite_start` and `os::test::junit::declare_suite_end`. These know exactly what test suite your in and can output that information. One less line of typing and you have similar output both in the console and in the report, wdyt?

stevekuznetsov replied:
It's an interesting thought. I think before `os::cmd` these types of declarations had value in that it was difficult to tell if forward progress was being made or not. What do we gain by keeping them at all now? Also, I decided to not add suite-specific information to  `os::test::junit:: declare_suite_end` as without a stack that gets hairy on reconciliation.

soltysh replied:
This can be applied as a followup, of course :)

soltysh commented on tools/junitreport/pkg/api/junit.xsd:
```diff
@@ -0,0 +1,203 @@
+         <xs:whiteSpace value="preserve" />
+      </xs:restriction>
+   </xs:simpleType>
+</xs:schema>
```
Now I remember why I hate xml+xsd, it's soooooo big :wink:

stevekuznetsov replied:
Not to mention completely unreadable by humans.

stevekuznetsov replied:
I guess I could refactor this to YAML as it's only used as a source of truth, not for processing, etc.

soltysh replied:
That would definitely be valuable :heart_eyes:

soltysh commented:
Still LGTM.

stevekuznetsov commented:
@deads2k @soltysh rebased, up-to-date again

deads2k commented:
post-1.2

@soltysh has the merge.

stevekuznetsov commented:
Tests succeeded but failed on the last step - publishing jUnit XML report. I think @smarterclayton turned that on. Otherwise, green.

smarterclayton commented:
Going to do this post 1.2 once I cut the branch.

stevekuznetsov commented:
re[test] for good luck

stevekuznetsov commented:
Flakes:
https://ci.openshift.redhat.com/jenkins/job/test_pull_requests_origin/14321/: https://github.com/openshift/origin/issues/7429
https://ci.openshift.redhat.com/jenkins/job/test_pull_requests_origin/14320: https://github.com/openshift/origin/issues/8433

re[test]

smarterclayton commented:
[test]

On Fri, Apr 8, 2016 at 5:25 PM, OpenShift Bot <notifications@github.com>
wrote:

> continuous-integration/openshift-jenkins/test FAILURE (
> https://ci.openshift.redhat.com/jenkins/job/test_pr_origin/2862/)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/6574#issuecomment-207611732>
>

stevekuznetsov commented:
conformance flakes on OOM issue: https://github.com/kubernetes/kubernetes/issues/23894
integration flake on `TestNewAppRunBuilds`:
```
--- FAIL: TestNewAppRunBuilds-2 (3.15s)
	newapp_test.go:1198: successful build from dockerfile with --to: unexpected error: no match for "openshift/origin-base"
```

re[test]

stevekuznetsov commented:
@soltysh ping

soltysh commented:
@stevekuznetsov still waiting for 1.2 cut

smarterclayton commented:
This is approved for master before 1.2 because it is test infrastructure.

soltysh commented:
Upon approval:
[merge]
